src/main/java/org/radarbase/appserver/service/storage/StoragePath.java

```diff
@@ -0,0 +1,169 @@
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+
+        private transient String pathPrefix = "";
+        private transient String file = "";
+        private transient boolean doCollectPerDay = false;
+        private transient String project = "";
+        private transient String subject = "";
+        private transient String topic = "";
+        private transient String folderPattern = "yyyyMMdd";
+        private transient String filePattern = "yyyyMMddHHmmss";
+        private transient String dirSep = "/";
+
+        public Builder filename(String filename) {
+            this.file = filename;
+            return this;
+        }
+
+        public Builder prefix(String prefix) {
+            Assert.notNull(prefix, "Prefix must not be null");
+            this.pathPrefix = prefix;
+            return this;
+        }
+
+        public Builder collectPerDay(boolean collectPerDay) {
+            this.doCollectPerDay = collectPerDay;
+            return this;
+        }
+
+        public Builder projectId(String projectId) {
+            Assert.notNull(projectId, "Project Id must not be null");
+            this.project = projectId;
+            return this;
+        }
+
+        public Builder subjectId(String subjectId) {
+            Assert.notNull(subjectId, "Subject Id must not be null");
+            this.subject = subjectId;
+            return this;
+        }
+
+        public Builder topicId(String topicId) {
+            Assert.notNull(topicId, "Topic Id must not be null");
+            this.topic = topicId;
+            return this;
+        }
+
+        public Builder dayFolderPattern(String dayFolderPattern) {
+            Assert.notNull(dayFolderPattern, "Day folder pattern must not be null");
+            this.folderPattern = dayFolderPattern;
+            return this;
+        }
+
+        public Builder fileTimestampPattern(String fileTimestampPattern) {
+            Assert.notNull(fileTimestampPattern, "File timestamp pattern must not be null");
+            this.filePattern = fileTimestampPattern;
+            return this;
+        }
+
+        public StoragePath build() {
+            Assert.isTrue(!file.isBlank(), "Filename must be set.");
+            Assert.isTrue(!project.isBlank(), "Project Id must be set.");
+            Assert.isTrue(!subject.isBlank(), "Subject Id must be set.");
+            Assert.isTrue(!topic.isBlank(), "Topic Id must be set.");
+
+            String pathInTopicDir = Stream.of(
+                    this.doCollectPerDay ? getDayFolder() : "",
+                    // Storing files under their original filename is a security risk, as it can be used to
+                    // overwrite existing files. We generate a random filename server-side to mitigate this risk.
+                    // See https://owasp.org/www-community/vulnerabilities/Unrestricted_File_Upload
+                    generateRandomFilename(this.file)
```

Comment:
Would it be possible to take out this method and add to a var then put the comment above that? For readability.

Reply:
Done!